x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PreviewDatafeedAction.java

```diff
@@ -87,7 +143,53 @@ public boolean equals(Object obj) {
                 return false;
             }
             Request other = (Request) obj;
-            return Objects.equals(datafeedId, other.datafeedId);
+            return Objects.equals(datafeedId, other.datafeedId)
+                && Objects.equals(datafeedConfig, other.datafeedConfig)
+                && Objects.equals(jobConfig, other.jobConfig);
+        }
+
+        public static class Builder {
+            private String datafeedId;
+            private DatafeedConfig.Builder datafeedBuilder;
+            private Job.Builder jobBuilder;
+
+            public Builder setDatafeedId(String datafeedId) {
+                this.datafeedId = datafeedId;
+                return this;
+            }
+
+            public Builder setDatafeedBuilder(DatafeedConfig.Builder datafeedBuilder) {
+                this.datafeedBuilder = datafeedBuilder;
+                return this;
+            }
+
+            public Builder setJobBuilder(Job.Builder jobBuilder) {
+                this.jobBuilder = jobBuilder;
+                return this;
+            }
+
+            public Request build() {
+                if (datafeedBuilder != null) {
+                    datafeedBuilder.setId("preview_id");
+                    if (datafeedBuilder.getJobId() == null && jobBuilder == null) {
+                        throw new IllegalArgumentException("[datafeed_config.job_id] must be set or a [job_config] must be provided");
+                    }
+                    if (datafeedBuilder.getJobId() == null) {
+                        datafeedBuilder.setJobId("preview_job_id");
+                    }
+                }
+                if (jobBuilder != null) {
```

Comment:
Is it valid to have a jobBuilder but no datafeedBuilder a null datafeedId? i.e. just the job config

Reply:
> Is it valid to have a jobBuilder but no datafeedBuilder a null datafeedId?

no, it isn't. The API MUST have either a datafeedBuilder or a datafeedId. The restrictive ctors ensure this (with null checks).